File: 2018/round1a/b.js
var readline = require('readline');

var rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});
var lines = [];
rl.on('line', function(input) {
    lines.push(input);
});
rl.on('close', function() {
    var t = parseInt(lines[0]);
    var l = 1;
    for (var i = 0; i < t; i++) {
        var tokens = lines[l++].split(' ').map((x) => +x);
        console.log('Case #%d: %s', i + 1, solve(tokens[0], tokens[1], tokens[2],
                lines.slice(l, l + tokens[2]).map((str) => str.split(' ').map((x) => +x))));
        l += tokens[2];
    }
});

function solve(r, b, c, cs) {
    cs = cs.map((x) => ({
        m: x[0],
        s: x[1],
        p: x[2]
    }));

    var min = 0;
    var max = b * Math.max.apply(Math, cs.map((x) => x.s))
        + Math.max.apply(Math, cs.map((x) => x.p));
    while (min < max) {
        var middle = Math.floor((min + max) / 2);
        if (check(middle, cs, r, b)) {
            max = middle;
        } else {
            if (min + 1 == max) break;
            min = middle;
        }
    }

    return max;
}
function check(t, cs, r, b) {
    cs.sort((c1, c2) => {
        return getCap(t, c2) - getCap(t, c1);
    });

    return b <= cs.slice(0, r).reduce((s, item) => s += getCap(t, item), 0);
}
function getCap(t, item) {
    return Math.max(0, Math.min(item.m, Math.floor((t - item.p) / item.s)));
}

File: 2018/round1a/b.py
import math

def getCap(t, item):
    [m, s, p] = item
    return max(0, min(m, math.floor((float(t) - p) / s)))

def check(t, cs, r, b):
    cs = sorted(cs, lambda c1, c2: int(getCap(t, c2) - getCap(t, c1)))
    return b <= sum(map(lambda item: getCap(t, item), cs[0:r]))

def solve(r, b, c, cs):
    ms = max(cs, key=lambda item: item[1])[1]
    mp = max(cs, key=lambda item: item[2])[2]

    tmin = 0
    tmax = b * ms + mp
    while tmin < tmax:
        middle = long(math.floor((tmin + tmax) / 2))
        if check(middle, cs, r, b):
            tmax = middle
        else:
            if tmin + 1 == tmax:
                break
            tmin = middle
    return tmax

t = int(raw_input())
for i in xrange(1, t + 1):
    r, b, c = map(int, raw_input().split(' '))

    cs = []
    for j in xrange(c):
        m, s, p = map(int, raw_input().split(' '))
        cs.append([m, s, p])

    print 'Case #{}: {}'.format(i, solve(r, b, c, cs))
